test/e2e/storage/non_csi_volumes.go

```diff
@@ -0,0 +1,1567 @@
+						testResource.setup(driver, curTestPattern)
+						testInput = testCase.createTestInput(curTestPattern, testResource)
+					})
+
+					AfterEach(func() {
+						testResource.cleanup()
+						driver.cleanupDriver()
+					})
+
+					curTestFunc(&testInput)
+				})
+			}
+		}
+	}
+})
+
+type testDriver interface {
+	createDriver()
+	cleanupDriver()
+	getDriverInfo() driverInfo
+}
+
+type inlineVolumeTestDriver interface {
+	testDriver
+	getVolumeSource(readOnly bool, fsType string) *v1.VolumeSource
+}
+
+type staticPVTestDriver interface {
+	testDriver
+	getPersistentVolumeSource(readOnly bool, fsType string) *v1.PersistentVolumeSource
+}
+
+type dynamicPVTestDriver interface {
+	testDriver
+	getDynamicProvisionStorageClass(fsType string) *storagev1.StorageClass
+}
+
+type driverInfo struct {
+	name                      string
+	fsGroup                   *int64
+	podSec                    v1.PodSecurityContext
+	privilegedSecurityContext bool
+	maxFileSize               int64
+
+	// supportablity for driver
+	supportedFsType  map[string]bool
+	isBlockSupported bool
+
+	// volumes test specific parameters
+	expectedContent string
+	needInjectHTML  bool
+
+	// volumeIO test specific parameter
+	testFile string
+
+	f      *framework.Framework
+	config framework.VolumeTestConfig
+	node   v1.Node
+}
+
+type testCase interface {
+	getTestPatterns() map[string]testPattern
+	initTestResource() testResource
+	createTestInput(testPattern, testResource) testInput
+	getTestFunc() func(*testInput)
+}
+
+type testVolType string
+
+var (
+	inlineVolume testVolType = "inlineVolume"
+	staticPV     testVolType = "staticPV"
+	dynamicPV    testVolType = "dynamicPV"
+)
+
+type testPattern struct {
+	testVolType testVolType
+	fsType      string
+	volBindMode storagev1.VolumeBindingMode
+	volMode     v1.PersistentVolumeMode
+}
+
+type testResource interface {
+	setup(testDriver, testPattern)
+	cleanup()
+}
+
+type testInput interface {
+	isTestInput() bool
+}
+
+func skipTestForDriver(testPattern testPattern, driver testDriver) {
+	driverInfo := driver.getDriverInfo()
+
+	// Check if testVolType is supported by driver
+	var isSupported bool
+	switch testPattern.testVolType {
+	case inlineVolume:
+		_, isSupported = driver.(inlineVolumeTestDriver)
+	case staticPV:
+		_, isSupported = driver.(staticPVTestDriver)
+	case dynamicPV:
+		_, isSupported = driver.(dynamicPVTestDriver)
+	default:
+		isSupported = false
+	}
+
+	if !isSupported {
+		framework.Skipf("Driver %s doesn't support %v -- skipping", driverInfo.name, testPattern.testVolType)
+	}
+}
+
+// Generic volume test resouce
+type genericVolumeTestResource struct {
+	driver    testDriver
+	volSource *v1.VolumeSource
+	pvc       *v1.PersistentVolumeClaim
+	pv        *v1.PersistentVolume
+	sc        *storagev1.StorageClass
+}
+
+func skipCreatingResourceForProvider(testPattern testPattern) {
+	fsType := testPattern.fsType
+
+	if fsType == "xfs" && framework.ProviderIs("gce", "gke") {
```

Comment:
Defined ```skipUnsupportedTest``` in ```testDriver``` and make it called from ```skipUnsupportedTest``` in test loop.

```skipUnsupportedTest``` become responsible for skipping other than FsType, like provider.
(```framework.SkipUnlessProviderIs``` is now moved from ```initDriver```, so it can be put outside BeforeEach.)